RFE: expose improved kernel logging features through libseccomp

tyhicks opened the pull request:
This pull request aims to provide expose the new seccomp logging functionality that's [on its way](https://lkml.kernel.org/r/<20170815220319.GA63342@beast>) into the upcoming 4.14 merge window (acked by Kees and in linux-next).

The main goal is to allow programs to use the new `SECCOMP_FILTER_FLAG_LOG` flag and `SECCOMP_RET_LOG` action to allow application developers more control over what seccomp actions are logged. To meet that goal, additional functionality had to be added to libseccomp to ensure that the new flag and action are only used when the currently running kernel supports them. This PR allows libseccomp to query the kernel to see if a given flag or action is supported. This is also important for the new `SECCOMP_RET_KILL_PROCESS` action that is also on its way into the upcoming 4.14 merge window.

A new libseccomp **global** attribute is created in order to allow applications, including the libseccomp test suite, to create filters containing a new action while running under an old kernel that doesn't support the new action. The attribute is global, rather than filter-specific, so that it can be set before `seccomp_init(3)` is called. While this functionality or something similar is necessary, it is a slight abuse of the existing `seccomp_attr_set(3)`/`seccomp_attr_get(3)` API resulting. [That design](https://groups.google.com/d/msg/libseccomp/mN7UjzrEjKI/Q8-amzxDBAAJ) excels in developer friendliness but lacks a bit in libseccomp code cleanliness while [another discussed design](https://groups.google.com/d/msg/libseccomp/mN7UjzrEjKI/ceX-6C1HBAAJ) is the exact opposite. The decision was made between Paul and myself to go with the former and that's what's present in this PR.

All the tests (C and Python) pass under "old" kernels as well as under a linux-next kernel containing the logging changes (linux-next tag `next-20170824`).

tyhicks commented:
I'll also mention that the method of filter flag validation used in 0fe4375 was [added as a kernel selftest](https://lkml.kernel.org/r/<1502426037-3777-5-git-send-email-tyhicks@canonical.com>) to ensure that the method of validation doesn't break in the future.

tyhicks commented:
It is best to review/merge #91 first so that the basic Python test changes in this PR are not skipped.

pcmoore commented on doc/man/man3/seccomp_action_valid.3:
```diff
@@ -0,0 +1,52 @@
+.B #include <seccomp.h>
+.sp
+.BI "int seccomp_action_valid(uint32_t " action ");"
+.sp
```
As we talked about over email, I'm not overly excited about public APIs like this; I would much prefer a "API level" check instead of having to check for specific things (e.g. actions, flags, etc.).

tyhicks replied:
Ok, lets see if we can find some time to chat about this at LSS over the next couple days.

IMO, this is really only needed for the libseccomp test suite (it could be handy for linking applications but definitely isn't necessary) so I hate to over-engineer it. On the other hand, I'm really not happy with the code that resulted from the global API approach as it feels pretty hackish to me.

tyhicks replied:
For anyone else following along, Paul and I were able to chat about this at LSS. He's planning to take a crack at the "API level" approach in the coming week or so and I'm going to rebase my SECCOMP_RET_LOG and SECCOMP_FILTER_FLAG_LOG changes on top of his patch(es).

pcmoore commented on include/seccomp.h.in:
```diff
@@ -65,6 +65,7 @@ enum scmp_filter_attr {
 	SCMP_FLTATR_CTL_TSYNC = 4,	/**< sync threads on filter load */
 	SCMP_FLTATR_API_TSKIP = 5,	/**< allow rules with a -1 syscall */
 	SCMP_GLBATR_CTL_KCHECKACTS = 6,	/**< ignore kernel support checks */
```
Ohh, GitHub is being sneaky/clever/stupid here ... I made the comment above in another commit and it is appearing here (presumably because this is the commit that introduces the code).

pcmoore commented on doc/man/man3/seccomp_action_valid.3:
```diff
@@ -22,7 +22,9 @@ The
-kernel that's currently running.
+kernel that's currently running. This function is affected by the
+.BR SCMP_GLBATR_CTL_KCHECKACTS
+attribute which will disable the kernel support check for newer actions.
```
This goes back to our email discussion about checking/setting a "API level" instead of checking/setting specific "things".  I'm not saying "no" to this approach yet, but I would like to see an "API level" approach before making a call on this.

tyhicks commented on src/system.h:
```diff
@@ -59,6 +59,7 @@ struct db_filter_col;
 #define SECCOMP_RET_TRAP	0x00030000U /* disallow and force a SIGSYS */
 #define SECCOMP_RET_ERRNO	0x00050000U /* returns an errno */
 #define SECCOMP_RET_TRACE	0x7ff00000U /* pass to a tracer or disallow */
+#define SECCOMP_RET_LOG		0x7ffc0000U /* allow after logging */
```
Right! Nice catch.

tyhicks commented on src/system.c:
```diff
@@ -101,6 +101,34 @@ int sys_chk_seccomp_syscall(void)
+ * otherwise.
+ */
+static int _sys_chk_seccomp_flag(int flag)
+{
```
The cod changed over time and the two functions were originally a bit more complicated than they ended up being. I agree that it now makes sense to merge them into the same function.

tyhicks commented on src/system.c:
```diff
@@ -98,6 +98,40 @@ int sys_chk_seccomp_syscall(void)
+ * If the action is supported one is returned, zero if unsupported, negative
+ * values on error. A return code of -EOPNOTSUPP indicates that the kernel is
+ * not new enough to support action queries.
+ */
```
That's fine with me.

tyhicks commented on src/system.c:
```diff
@@ -98,6 +98,40 @@ int sys_chk_seccomp_syscall(void)
+int sys_chk_seccomp_action(uint32_t action)
+{
+	int rc;
+
```
I agree and I believe that it is valid for the existing actions. What was it that makes you think it may not be valid for the existing actions?

pcmoore replied:
Looking at this again .... I have absolutely no idea, sorry about that.

tyhicks commented on src/system.c:
```diff
@@ -98,6 +98,40 @@ int sys_chk_seccomp_syscall(void)
+
+	rc = syscall(_nr_seccomp, SECCOMP_GET_ACTION_AVAIL, 0, &action);
+	if (rc == -1) {
+		if (errno == EINVAL) {
```
I didn't run it prior to submitting the PR (my bad!). However, I did just run it and the curly braces weren't flagged.

tyhicks replied:
In fact, the entire PR passed `make check-syntax`. Lucky me.

tyhicks commented on src/system.c:
```diff
@@ -101,6 +101,34 @@ int sys_chk_seccomp_syscall(void)
+	return 0;
+supported:
+	return 1;
+}
```
Heh, agreed. This is another case of the flag checking functions becoming more simplified as I fine-tuned the patches prior to the PR. I should have fine-tuned a little bit more. :)

pcmoore replied:
See my comments in the later patches, but it seems like we need the ability to *force* the flag value, e.g. `sys_set_seccomp_flag_kernel(...)`.

tyhicks commented on src/python/libseccomp.pxd:
```diff
@@ -72,6 +72,7 @@ cdef extern from "seccomp.h":
     cdef enum:
         SCMP_ACT_KILL
         SCMP_ACT_TRAP
+        SCMP_ACT_LOG
```
Yes and I will be submitting a PR against libseccomp-golang once the dust settles on this PR.

tyhicks commented on src/db.c:
```diff
@@ -886,6 +890,17 @@ int db_col_attr_set(struct db_filter_col *col,
+			/* supported */
+			rc = 0;
+			col->attr.log_enable = (value ? 1 : 0);
+		} else if (rc == 0) {
```
Yes, it does.

tyhicks commented:
@pcmoore I've made all of the changes there were requested, with the exception of the API level check that you planned to take a stab at, and pushed the new commits. The changes include:

- src/python/seccomp.pyx: Fix documentation typo
- src/system.c: Merge _sys_chk_seccomp_flag() into sys_chk_seccomp_flag()
  - Got rid of overly complicated unsupported and supported jump labels
- src/system.c: Only return 0 (unsupported) or 1 (supported) from sys_chk_seccomp_flag() 
- src/system.h: Ensure SECCOMP_RET_LOG is defined even when the system header is available

tyhicks commented:
I just noticed that I missed something. You asked for no negative values (only 0 for unsupported and 1 for supported) to be returned from sys_chk_seccomp_action() yet I only made that change for sys_check_seccomp_flag(). I assume that you want both functions to act similar to I'll retain the new behavior for sys_check_seccomp_flag() and push some new patches after I've updated sys_chk_seccomp_action().

tyhicks commented:
Alright, I've fixed up the return codes for sys_chk_seccomp_action().

tyhicks commented:
I've opened https://github.com/seccomp/libseccomp-golang/pull/29 for the corresponding changes needed in libseccomp-golang.

pcmoore commented:
@tyhicks my apologies on the late replies, I've been trying to get caught up from last week and OSS/LSS.

> I've made all of the changes there were requested, with the exception of the API level check that you planned to take a stab at ...

I put together a fairly crude patch to demonstrate what I was talking about, you can find it in the [working-api_level](https://github.com/seccomp/libseccomp/commits/working-api_level) branch.  It's incomplete, untested, etc. but at least it compiles ;)

Unfortunately I haven't had a chance yet to look at the changes you've made.

> You asked for no negative values (only 0 for unsupported and 1 for supported) to be returned from sys_chk_seccomp_action() yet I only made that change for sys_check_seccomp_flag(). I assume that you want both functions to act similar to I'll retain the new behavior for sys_check_seccomp_flag() ...

I need to go back and look at that code.  I realized when I was doing the API level stuff that the negative values may be useful, e.g. specifying a flag/action value that doesn't exist.  I apologize for the confusion.

pcmoore commented:
FYI, I finished the API level patch and merged it to master, can you rebase your PR on top of that so we can re-review it?

pcmoore commented:
Ooops, mean to include the commit ID for the API level commit, here it is: e89d18205c7dcd7582f41051cd6389c9b12dfccf

tyhicks commented:
> FYI, I finished the API level patch and merged it to master, can you rebase your PR on top of that so we can re-review it?

Thanks! Sure, I'll rebase, retest, and update this PR.

While rebasing, I was surprised by this:

`
+# NOTE: this is a NULL test since we don't support the seccomp_version() API
+#       via the libseccomp python bindings
`

What's the reasoning for that? The python tests (and Python client code) will need the seccomp_version() API to be exposed.

tyhicks commented:
> The Python tests (and Python client code) will need the seccomp_api_get() and seccomp_api_set() API to be exposed.

I've proposed a patch (https://github.com/seccomp/libseccomp/pull/98/commits/2bd18beda4ecb28a8c0f81e6ef821bd80276c7c7) that exposes the API level through the Python bindings in #98

tyhicks commented:
@pcmoore I've (locally) rebased the changes in this PR on top of the changes proposed in #98 in order to use the Python bindings work that I did in that PR. I guess I'll wait until you're able to review #98 before updating this PR but let me know if you want me to do something different.

pcmoore commented:
@tyhicks Thanks!  I'm swamped this week working towards a Friday, October 13th deadline, but I'll revisit this early next week.  I'm sorry for the delay, I appreciate your work and your patience.

tyhicks commented:
I've rebased onto the current state of #98, since we seem to be finalizing that PR, and I've updated this PR in the case that you are able to carve out enough time to review the rest of both PRs at once. I hope this doesn't complicate things...

tyhicks commented:
I did a quick rebase and push now that the API level fixups PR is merged. The commit hashes changed from what was in the PR to what was merged so a rebase of this PR was necessary.

pcmoore commented on src/api.c:
```diff
@@ -102,6 +102,10 @@ static unsigned int _seccomp_api_update(void)
 		level = 2;
 
+	/* level 3 */
+	if (sys_chk_seccomp_flag(SECCOMP_FILTER_FLAG_LOG) == 1)
```
The API level is a "everything and below" indicator, meaning that if we say we are operating at level 3 it means we support *everything* at level 3 and below.  For that reason I think this check should be something like the following:

    if ((level == 2) && (sys_chk_seccomp_flag(SECCOMP_FILTER_FLAG_LOG) == 1))

Granted, I think in most cases if FLAG_LOG is supported then everything else will be supported as well, but I would rather be safe.

tyhicks replied:
I agree. Nice catch!

pcmoore commented on src/api.c:
```diff
@@ -124,14 +128,17 @@ API const unsigned int seccomp_api_get(void)
-		break;
+	case 3:
+		sys_set_seccomp_flag(SECCOMP_FILTER_FLAG_LOG, true);
+	/* fallthrough */
```
Nitpicky, but since you're probably going to respin for the above comment, please indent this comment to the same level as the code :)

tyhicks replied:
This is actually what astyle requires. I originally had it indented like what you're asking for but astyle flagged it as incorrect.

pcmoore replied:
Huh, okay, sorry for the noise.

pcmoore commented on src/api.c:
```diff
@@ -124,14 +128,17 @@ API const unsigned int seccomp_api_get(void)
+	/* fallthrough */
 	case 2:
 		sys_set_seccomp_syscall(true);
 		sys_set_seccomp_flag(SECCOMP_FILTER_FLAG_TSYNC, true);
```
It seems as though we should force off the FLAG_LOG functionality in the case of API levels 1 and 2, yes?

tyhicks replied:
Hmm... you're right.

That means that the odd indentation of "fallthrough" doesn't matter because we can't fall through any longer.

pcmoore commented on src/system.h:
```diff
@@ -114,6 +117,8 @@ typedef struct sock_filter bpf_instr_raw;
 int sys_chk_seccomp_syscall(void);
 void sys_set_seccomp_syscall(bool enable);
 
+int sys_chk_seccomp_action(uint32_t action);
```
I'm also wondering if we need a `sys_set_seccomp_action(...)` function.

tyhicks replied:
We do need one. It gets added in the `all: add support for new log action` patch.

pcmoore replied:
Ah ha, found it later in the patchset.  It probably should be included here, but not the end of the world.

tyhicks replied:
It sounds like you're not *requiring* the patchset surgery needed to move `sys_set_seccomp_action()` up a patch in the series so I'm going to opt for leaving it where it is at rather than risk making a mistake since we seem to be close to merging this. Let me know if that's a problem.

pcmoore commented:
@tyhicks I just finished going through the latest patchset (comments inline), sort out those last few issues, or help me understand why I'm wrong, and I think we are all set.

tyhicks commented:
@pcmoore Thanks! I think your feedback is addressed in the latest push.

pcmoore commented:
Everything should now be merged (commit IDs below), thanks again @tyhicks!

0cc720376 tests: add test for SCMP_ACT_LOG of all syscalls
649d4fa3e tests: add SCMP_ACT_LOG test to 06-sim-actions
8f37ea87b tests: test suite infrastructure changes for SCMP_ACT_LOG
3b22b153c all: add support for new log action
b61042bf4 system: runtime check if an action is supported by the kernel
d0e11951f all: add support for new log filter flag
8a8576c9e system: runtime check if a filter flag is supported by the kernel

187sec commented:
Run setup daphcnal
